src/graphnet/data/pipeline.py

```diff
@@ -0,0 +1,173 @@
+from abc import ABC, abstractmethod
+from typing import List
+import numpy as np
+from abc import abstractmethod
+import torch
+import dill
+from graphnet.data.constants import TRUTH, FEATURES
+from graphnet.models.training.utils import get_predictions, make_dataloader
+from pytorch_lightning import Trainer
+from functools import reduce
+import pandas as pd
+import os
+import sqlalchemy
+import sqlite3
+
+class InSQLitePipeline(ABC):
+    """Creates a SQLite database with truth and GNN predictions and, if available, RETRO reconstructions. Made for analysis.
+    """
+    def __init__(self, module_dict, features, truth, device, retro_table_name = 'retro',outdir = None,batch_size = 100, n_workers = 10, pipeline_name = 'pipeline'):
+        """Initializes the pipeline
+
+        Args:
+            module_dict (dict): A dictionary with GNN modules from GraphNet. E.g. {'energy': gnn_module_for_energy_regression}
+            features (list): list of input features for the GNN modules
+            truth (list): list of truth for the GNN ModuleList
+            device (torch._device): the device used for computation
+            retro_table_name (str, optional): Name of the retro table for. Defaults to 'retro'.
+            outdir (path, optional): the directory in which the pipeline database will be stored. Defaults to None.
+            batch_size (int, optional): batch size for inference. Defaults to 100.
+            n_workers (int, optional): number of workers used in dataloading. Defaults to 10.
+            pipeline_name (str, optional): name of the pipeline. If such a pipeline already exists, an error will be prompted to avoid overwriting. Defaults to 'pipeline'.
+        """
+        self._pipeline_name = pipeline_name
+        self._device = device
+        self.n_workers =  n_workers
+        self._features = features
+        self._truth =  truth
+        self._batch_size = batch_size
+        self._outdir = outdir
+        self._module_dict = module_dict
+        self._retro_table_name = retro_table_name
+
+    def __call__(self, database, pulsemap) -> dict:
+        """Runs inference of each field in self._module_dict[target]['']
+
+        Args:
+            database (path): path to database with pulsemap and truth
+            pulsemap (str): name of pulsemaps
+        """
+        outdir = self._get_outdir(database)
+        if isinstance(self._device, str): # Because pytorch lightning insists on breaking pytorch cuda device naming scheme
+            device = int(self._device[-1])
+        if os.path.isdir(outdir) == False:
+            dataframes = []
+            for target in self._module_dict.keys():
+                dataloader = self._make_dataloader(database, pulsemap)
```

Comment:
Currently, the dataloader is being constructed once for every model? Is this necessary?

Reply:
We could set `persistent_workers=True` if that's the issue?

Reply:
its an iterable so by creating an identical one for every model makes me sleep well at night.